Fix receiving custom snapshot properties

loli10K opened the pull request:
We never initialize `snapprops_nvlist` in `zfs_receive_one`, so custom properties aren't applied to the received snapshot.

Should fix https://github.com/zfsonlinux/zfs/issues/4338, more info in comments.

behlendorf commented:
@loli10K thanks for digging in to this and opening a pull request.  I won't be able to look at this carefully until tomorrow but in the meanwhile it would be great if you could added a simple test case for sending user properties.  To my surprise I didn't see one in `tests/zfs-tests/tests/functional/cli_root/zfs_send/` directory.   I'd suggest adding a `zfs_send_008_pos.ksh` to test this, you can model it on one of the existing test cases.  You'll also want to make sure you add it to the `linux.run` file.

@dweeezil can you look at this.  Your previous commit dropped this hunk e890dd85a7522730ad46daf68150aafd3952d0c1.

loli10K commented:
@behlendorf the commit message is a typo of mine, the issue is actually on the receiving side: should i add a `zfs_receive_XXX_pos.ksh` instead?

behlendorf commented:
@loli10K adding it as a recv test as you've done is fine.  Thanks for adding the test case.

loli10K commented:
Adding new test cases revealed that the initial fix is not working correctly.

Without the first commit (initialize `snapprops_nvlist` in `zfs_receive_one`) ZFS is able to receive a full send stream and set its properties on the dataset, but fails to receive the snapshot properties.

With the "fix" commit ZFS is able to receive the snapshot properties but instead fails to receive the dataset properties with: `cannot receive snap1 property on testpool: invalid property value`.

Back to the debugger.

loli10K commented:
Update: the problem with my initial patch is the incompatibility with the implementation of `zfs_ioc_recv_new` (https://github.com/zfsonlinux/zfs/commit/43e52eddb13d8accbd052fac9a242ce979531aa4): we don't pass the list of properties to be received via a `zfs_cmd_t` anymore, we send a `nvlist_t` directly via the new function `lzc_receive_one`.

Workflow before https://github.com/zfsonlinux/zfs/commit/43e52eddb13d8accbd052fac9a242ce979531aa4:

1. get a new `zfs_cmd_t zc`
2. update it with the dataset properties: `zcmd_write_src_nvlist(hdl, &zc, props)`
3. lookup the snapshot properties: `nvlist_lookup_nvlist(fs, "snapprops", &props)`. (_this is the code i added and tested initially on my systems (0.6.5.X) but breaks 0.7.0_)
4. receive the send stream: `zfs_ioctl(hdl, ZFS_IOC_RECV, &zc)`
5. update the snapshot properties only if `snapprops_nvlist` isn't NULL

Everything seems to work with the previous code. Now with the new `zfs_ioc_recv_new` implementation:

1. `zfs_cmd_t zc` is not needed anymore
2. we don't update it with: `zcmd_write_src_nvlist(hdl, &zc, props)`
3. lookup the snapshot properties: `nvlist_lookup_nvlist(fs, "snapprops", &props)`. (_this is the code i added and tested initially on my systems (0.6.5.X) but breaks 0.7.0_)
4. receive the send stream: `lzc_receive_one(.., props, ..., &prop_errors)`
5. update the snapshot properties only if `snapprops_nvlist` isn't NULL

This patch won't let ZFS receive the dataset properties because it updates `props` directly with the value of the snapshot nvlist before `lzc_receive_one(.., props, ..., &prop_errors)`. Indeed the error reported is `cannot receive ${snapshotname} property on testpool: invalid property value`. In the old code we could "save" the dataset properties to be received in a `zfs_cmd_t` before looking up the snapshot properties.

I just applied the same patch on the commit just before https://github.com/zfsonlinux/zfs/commit/43e52eddb13d8accbd052fac9a242ce979531aa4 and it seems to work.

behlendorf commented on lib/libzfs/libzfs_sendrecv.c:
```diff
@@ -3122,6 +3122,11 @@ zfs_receive_one(libzfs_handle_t *hdl, int infd, const char *tosnap,
 			    zfs_prop_to_name(ZFS_PROP_CANMOUNT), 0));
 		}
+		if (0 == nvlist_lookup_nvlist(fs, "snapprops",
+			&snapprops_nvlist)) {
```
It would make sense to introduce a local variable in this block simply to be used to lookup the snapprops nvlist.  As was done with `props` in the original version of this function.  That way `snapprops_nvlist` will only be non-NULL when snapname exists.

loli10K replied:
You're right about the non-NULL condition, i just thought that if the first `nvlist_lookup_nvlist` doesn't exit with errors and incorrectly sets `snapprops_nvlist` then we either find `snapname` in the second lookup (overwriting the nvlist with the correct value) or we fail the `VERIFY()`.

behlendorf commented on tests/zfs-tests/tests/functional/cli_root/zfs_receive/zfs_receive_013_pos.ksh:
```diff
@@ -0,0 +1,83 @@
+#
+
+#
+# Copyright 2016, [name of copyright owner]. All rights reserved.
```
You wrote the test case, feel free to add your name here.  Or your companies name if that's more appropriate.

behlendorf commented on lib/libzfs/libzfs_sendrecv.c:
```diff
@@ -3090,6 +3090,7 @@ zfs_receive_one(libzfs_handle_t *hdl, int infd, const char *tosnap,
 	uint64_t errflags = 0;
 	uint64_t parent_snapguid = 0;
 	prop_changelist_t *clp = NULL;
+	nvlist_t *lookup_nvlist = NULL;
```
I'd suggest moving this in to the `if (stream_avl != NULL)` block.

loli10K replied:
Done, and rebased on current master

behlendorf commented on lib/libzfs/libzfs_sendrecv.c:
```diff
@@ -3110,6 +3110,8 @@ zfs_receive_one(libzfs_handle_t *hdl, int infd, const char *tosnap,
 
 	if (stream_avl != NULL) {
+		nvlist_t *lookup_nvlist = NULL;
+
```
[style] No need for an empty line here.

behlendorf commented on lib/libzfs/libzfs_sendrecv.c:
```diff
@@ -3125,6 +3127,11 @@ zfs_receive_one(libzfs_handle_t *hdl, int infd, const char *tosnap,
 			    zfs_prop_to_name(ZFS_PROP_CANMOUNT), 0));
 		}
+		if (0 == nvlist_lookup_nvlist(fs, "snapprops",
+			&lookup_nvlist)) {
```
[style] Four space indent here for continued conditional.

```
+		if (0 == nvlist_lookup_nvlist(fs, "snapprops",
+		    &lookup_nvlist)) {
```

Or better yet if you picked a shorter variable name this could all be done on one line.  Since the scope of this variable is so small it doesn't need to be super descriptive.

```
                if (0 == nvlist_lookup_nvlist(fs, "snapprops", &lookup)) {
```

loli10K commented:
Sorry for all the trouble, i will try to keep these things in mind for the next PR.

behlendorf commented:
@loli10K thanks for digging in to this issue.
